### PR TITLE
[CUDA] Elide dense causal masks in long-context prefill

### PR DESCRIPTION
**Pull request by Gasoonjia**
## Summary

Elide causal-mask-elision work on top of `split_k_nan`.

- Generate bottom-right causal bounds inside Triton SDPA instead of materializing dense masks.
- Cover standard SDPA plus the Gemma4 and Muse-Glimmer source transformations; TQ4 SDPA and its operator schema remain unchanged.
- Preserve explicit arbitrary masks by composing them with the reconstructed causal bound.
- Apply bottom-right causal masking in the verifier-sized (`L_q=2..4`) split-K path.
- Preserve the numerically stable online-softmax implementation from `split_k_nan`; this does not restore the previous `phi=5` approximation.

## Operator-level performance

RTX 5090, BF16 GQA (`B=1, Hq=32, Hkv=2, D=128`), explicit dense causal mask versus in-kernel causal reconstruction:

| Lq | Lkv | Dense mask | In-kernel causal | Speedup |
|---:|---:|---:|---:|---:|
| 128 | 16,384 | 0.643 ms | 0.505 ms | 1.27x |
| 512 | 16,384 | 1.190 ms | 1.008 ms | 1.18x |
| 512 | 32,768 | 2.397 ms | 2.004 ms | 1.20x |

## Test plan

- `lintrunner init && lintrunner -a` (clean).
- Standard Triton SDPA: 27 passed.
- TQ4 SDPA schema and implementation match `split_k_nan`; focused long-context checks: 2 passed.
- Full TQ4 kernel suite: 35 passed, 2 skipped; the export-only case requires the generated `program.fbs` resource unavailable in the ad-hoc GPU worktree.
- `git diff --check`.

### CODE REVIEW

File: backends/cuda/triton/kernels/sdpa.py [hunk 1]
```diff
@@ -1007,6 +1039,13 @@ def sdpa(
     else:
         kv_len_t = None
 
+    mask_is_causal = is_causal and HAS_KV_LEN
+    if mask_is_causal:
```
Comment (shoumikhin) [inline backends/cuda/triton/kernels/sdpa.py]: This clears the dense mask, but the small-query split-K launcher below never learns the mask was causal. `_launch_small_query_splitk` has no `mask_is_causal` parameter and no causal branch; only `_launch_pow2_kernel` and `_launch_non_pow2_kernel` were threaded. So for `2 <= L_q <= 4` with `kv_len` set, a power-of-two head dim and `L_kv >= 256`, the output comes out fully non-causal. I ran it on an H100: it matches a non-causal reference to bf16 noise, while the same shapes on the merge base are correct.

That window is exported. `export_dflash.py` uses `Dim("target_seq_len", min=1, max=4)`, and the runtime verify length is `min(n_draft + 1, room)`, which lands in the broken range on every speculative step with a nonzero draft.

Worth noting the public `sdpa_small_query_splitk` op already refuses `L_q != 1` with a `torch._check`, and there is a test asserting the refusal. `torch.ops.triton.sdpa` reaches the launcher directly and skips that guard.

**Review by shoumikhin — COMMENTED** [review]
Went through this one carefully, since the mask elision touches every SDPA launch path.

Two things look blocking to me, both inline: the 2 to 4 query-length window loses causal masking completely, and the TurboQuant call site under `examples/models/muse-glimmer/` was not migrated, so it now raises. Two smaller ones inline as well: an explicit mask is dropped rather than composed, and the new `TARGETS` file hides the existing `BUCK`.

The rest reads well. The per-tile causal bound and the bottom-right alignment look right on the paths that were threaded, and the two new causal tests do exercise them.

File: backends/cuda/triton/kernels/sdpa.py [hunk 2]
```diff
@@ -1007,6 +1039,13 @@ def sdpa(
...
+        HAS_MASK = False
```
Comment (shoumikhin) [inline backends/cuda/triton/kernels/sdpa.py]: An explicit `attn_mask` is dropped here rather than composed with the causal bound. With both a mask and `is_causal` plus `kv_len`, I get output that is bit identical to passing no mask at all, and different from the mask-only result. The merge base applied both. Padding masks, sliding windows and blocked keys would vanish with no warning.

The PR description says explicit mask behaviour is preserved, so one of the two needs to change: either compose them, or raise on the combination.

File: backends/cuda/triton/kernels/tq4_sdpa.py [hunk 3]
```diff
@@ -575,7 +575,6 @@ def tq4_sdpa(
     is_causal: bool = False,
     scale: Optional[float] = None,
     kv_len: Optional[torch.Tensor] = None,
-    mask_is_causal: bool = False,
 ) -> torch.Tensor:
```
Comment (shoumikhin) [inline backends/cuda/triton/kernels/tq4_sdpa.py:579]: Dropping this parameter from the registered schema has two side effects.

`examples/models/muse-glimmer/source_transformations/cuda.py:463` still passes 12 arguments to `triton::tq4_sdpa` and now raises `expected at most 11 argument(s) but received 12`. This commit already edits that file, just a different function; the docstring at `:426` is stale too. One warning on the fix: deleting the extra argument on its own leaves `is_causal=False` and the layer goes silently non-causal (about 69% relative change when I tried it). It needs `is_causal=True`, the same treatment gemma4 got.

Second, a `.pt2` saved before this change still loads afterwards, the trailing `True` is silently dropped, and the model becomes non-causal. This file already handles that exact case the other way round: `phi` is kept with the note "deprecated, accepted for operator-schema compatibility, and ignored". Keeping `mask_is_causal` as an ignored argument would avoid the silent rewrite.

File: backends/cuda/tests/TARGETS [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+    deps = [
+        "//caffe2:torch",
+        "//executorch/backends/cuda:cuda_partitioner",
+        "//executorch/backends/cuda:cuda_backend",
+        "//executorch/exir:lib",
+        "//executorch/exir/backend:compile_spec_schema",
+    ],
+)
+
+# TEMP: mask-dev validation of triton.sdpa mask_is_causal (remove after run)
```
Comment (shoumikhin) [inline backends/cuda/tests/TARGETS]: Two problems with this file.

The target is labelled "remove after run" and names the parameter this same PR deletes, so I do not think it should land at all.

The bigger one: adding `TARGETS` here shadows the existing `BUCK`. `.buckconfig` sets `name_v2 = TARGETS,BUCK`, and buck2 takes the first candidate present, so `BUCK` stops being parsed. `BUCK` is the only file that loads `targets.bzl`, and `targets.bzl` holds the only definition of `test_triton_sdpa_splitk`, which is the target for the test file the next PR in this stack modifies. That target quietly disappears. The new file also skips the `is_fbcode` guard that `targets.bzl` applies to everything. And 8032e6982e ("Rename build files from TARGETS to BUCK") removed this exact file on purpose.

Suggest deleting it, and adding any lasting target to `targets.bzl` instead.

**Comment by digantdesai** [comment]
merged split-k-nan FYI

File: backends/cuda/triton/kernels/sdpa.py (hunk 1 above)
Comment (Gasoonjia) [reply backends/cuda/triton/kernels/sdpa.py]: Fixed. The small-query split-K kernel now receives an explicit causal flag and applies the bottom-right causal bound for L_q=2..4. I also added regression coverage for the exported verifier-sized path.

File: backends/cuda/triton/kernels/sdpa.py (hunk 2 above)
Comment (Gasoonjia) [reply backends/cuda/triton/kernels/sdpa.py]: Fixed. An explicit attention mask is now preserved and composed with the reconstructed causal bound instead of being cleared. The combined attn_mask + is_causal + kv_len case is covered by a regression test.

File: backends/cuda/triton/kernels/tq4_sdpa.py (hunk 3 above)
Comment (Gasoonjia) [reply backends/cuda/triton/kernels/tq4_sdpa.py:579]: Agreed. I reverted all TQ4 SDPA changes from this PR, including its schema and tests. TQ4 is now unchanged from split_k_nan; future external KV-cache work will remove the need for a separate TQ4 attention path.

File: backends/cuda/tests/TARGETS (hunk 4 above)
Comment (Gasoonjia) [reply backends/cuda/tests/TARGETS]: Fixed by removing the temporary TARGETS file entirely. The existing BUCK file and targets.bzl definitions, including the is_fbcode guard and split-K test target, are preserved.

**Review by digantdesai — APPROVED** [review]
